### PR TITLE
[Feature] Support for JSON `minimum` , `maximum`, `exclusiveMinimum`, and `exclusiveMaximum`

### LINKED ISSUE #1019: support for minimum/maximum in JSON schemas
Currently we don't support the numerical bounds for JSON numbers. I wrote some code that compiles the bounds into regular expressions.

I don't have time right now to make it into full PR right now - maybe someone else wants to take a look.

There are two functions in the file, one for integer and one for real (float) ranges. It only supports inclusive ranges right now (not a problem for integers, but for floats would need some more work to support exclusive).

```python
import re
import math


def mk_or(parts: list[str]) -> str:
    if len(parts) == 1:
        return parts[0]
    return "(" + "|".join(parts) + ")"


def num_digits(n: int) -> int:
    return len(str(n))


def rx_int_range(left: int, right: int) -> str:
    assert left <= right
    if left < 0:
        if right < 0:
            return "(-" + rx_int_range(-right, -left) + ")"
        else:
            return "(-" + rx_int_range(0, -left) + "|" + rx_int_range(0, right) + ")"
    else:
        if num_digits(left) == num_digits(right):
            l = str(left)
            r = str(right)
            if left == right:
                return f"({l})"

            lpref = l[:-1]
            lx = l[-1]
            rpref = r[:-1]
            rx = r[-1]

            # 1723-1728 => 172[3-8]
            if lpref == rpref:
                return f"({lpref}[{lx}-{rx}])"

            # general case 723-915 => 72[3-9]|(73-90)[0-9]|91[0-5]
            left_rec = int(lpref)
            right_rec = int(rpref)
            assert left_rec < right_rec
            parts = []

            # optimizations:
            # for 0, we have 720-915 => (72-90)[0-9]|91[0-5]
            if lx != "0":
                left_rec += 1
                parts.append(f"{lpref}[{lx}-9]")
            # for 9 we have 723-919 => 72[3-9]|(73-91)[0-9]
            if rx != "9":
                right_rec -= 1
                parts.append(f"{rpref}[0-{rx}]")

            # the middle can be empty 723-734 => 72[3-9]|73[0-4]
            if left_rec <= right_rec:
                inner = rx_int_range(left_rec, right_rec)
                parts.append(f"{inner}[0-9]")

            return mk_or(parts)
        else:
            break_point = 10 ** num_digits(left) - 1
            return mk_or(
                [rx_int_range(left, break_point), rx_int_range(break_point + 1, right)]
            )


def lexi_x_to_9(x: str) -> str:
    if x == "":
        return "[0-9]*"
    if len(x) == 1:
        return f"[{x}-9][0-9]*"
    x0 = int(x[0])
    parts = [x[0] + lexi_x_to_9(x[1:])]
    if x0 < 9:
        parts.append(f"[{x0 + 1}-9][0-9]*")
    return mk_or(parts)


def lexi_0_to_x(x: str) -> str:
    if x == "":
        return ""  # don't allow trailing zeros
    x0 = int(x[0])
    parts = [x[0] + lexi_0_to_x(x[1:])]
    if x0 > 0:
        parts.append(f"[0-{x0 - 1}][0-9]*")
    return mk_or(parts)


def lexi_range(ld: str, rd: str) -> str:
    assert len(ld) == len(rd)
    if ld == rd:
        return ld
    l0 = int(ld[0])
    r0 = int(rd[0])
    # common prefix: 137-144 => 1(37-44)
    if r0 == l0:
        return ld[0] + lexi_range(ld[1:], rd[1:])
    assert l0 < r0
    # 23470-82142 => 2(347-999)|[3-7][0-9]*|8(0000-2142)
    parts = [ld[0] + lexi_x_to_9(ld[1:].rstrip("0"))]
    # is the [3-7][0-9]* part empty?
    if l0 + 1 < r0:
        parts.append(f"[{l0 + 1}-{r0 - 1}][0-9]*")
    parts.append(rd[0] + lexi_0_to_x(rd[1:].rstrip("0")))
    return mk_or(parts)

def float_to_str(f: float) -> str:
    s = f"{f:f}"
    return s.rstrip("0").rstrip(".")

def rx_float_range(left: float, right: float) -> str:
    assert left <= right
    if left < 0:
        if right < 0:
            return "(-" + rx_float_range(-right, -left) + ")"
        else:
            return (
                "(-"
                + rx_float_range(0.0, -left)
                + "|"
                + rx_float_range(0.0, right)
                + ")"
            )
    else:
        l = float_to_str(left)
        r = float_to_str(right)
        if left == right:
            return f"({l})"

        if "e" in l or "e" in r:
            raise ValueError("Scientific notation not supported")
        if not math.isfinite(left) or not math.isfinite(right):
            raise ValueError("Infinite numbers not supported")

        left_rec = int(l.split(".")[0])
        right_rec = int(r.split(".")[0])

        ld = (l.split(".") + [""])[1]
        rd = (r.split(".") + [""])[1]

        # 17.123-17.1448 -> 17.((1230-1447)[0-9]*|1448)
        if left_rec == right_rec:
            while len(ld) < len(rd):
                ld += "0"
            while len(rd) < len(ld):
                rd += "0"
            suff = "\\." + lexi_range(ld, rd)
            if int(ld) == 0:
                return f"({left_rec}({suff})?)"
            else:
                return f"({left_rec}{suff})"

        parts = []

        # 7.321-22.123 -> 7.(321-999)|8-21(.[0-9]+)?|22.(000-123)
        if ld:
            parts.append(f"({left_rec}\\.{lexi_x_to_9(ld)})")
            left_rec += 1

        if right_rec - 1 >= left_rec:
            inner = rx_int_range(left_rec, right_rec - 1)
            parts.append(f"({inner}(\\.[0-9]+)?)")

        if rd:
            parts.append(f"({right_rec}(\\.{lexi_0_to_x(rd)})?)")
        else:
            parts.append(f"{right_rec}")

        return mk_or(parts)


def do_test_int_range(rx: str, left: int, right: int) -> None:
    print(rx)
    n = left - 1000
    while n < right + 1000:
        m = re.fullmatch(rx, str(n)) is not None
        f = left <= n <= right
        if m != f:
            print(f"Failed for {n} match={m} expected={f}")
            assert False
        n += 1


def test_int_range(left: int, right: int) -> None:
    print(f"Testing range {left}-{right}")
    rx = rx_int_range(left, right)
    do_test_int_range(rx, left, right)


def test_float_range(left: float, right: float) -> None:
    print(f"Testing range {left}-{right}")
    rx = rx_float_range(left, right)
    do_test_int_range(rx, math.ceil(left), math.floor(right))
    eps = 0.000001
    for x in [left, right, 0, int(left), int(right)]:
        for off in [0, -eps, eps, 1, -1]:
            n = x + off
            ns = float_to_str(n)
            m = re.fullmatch(rx, ns) is not None
            f = left <= n <= right
            if m != f:
                print(f"Failed float for {ns} match={m} expected={f}")
                assert False


def main():
    test_int_range(0, 9)
    test_int_range(1, 7)
    test_int_range(0, 99)
    test_int_range(13, 170)
    test_int_range(13, 17)
    test_int_range(13, 27)
    test_int_range(13, 57)
    test_int_range(72, 91)
    test_int_range(723, 915)
    test_int_range(23, 915)
    test_int_range(-1, 915)
    test_int_range(-9, 9)
    test_int_range(-3, 3)
    test_int_range(-3, 0)
    test_int_range(-72, 13)

    test_float_range(0, 10)
    test_float_range(-10, 0)

    test_float_range(0.5, 0.72)
    test_float_range(0.5, 1.72)
    test_float_range(0.5, 1.32)
    test_float_range(0.3245, 0.325)
    test_float_range(1, 2.34)
    test_float_range(1.33, 2)
    test_float_range(1, 10.34)
    test_float_range(1.33, 10)
    test_float_range(-1.33, 10)
    test_float_range(-17.23, -1.33)
    test_float_range(-1.23, -1.221)
    test_float_range(-10.2, 45293.9)


if __name__ == "__main__":
    main()
```

### PR DESCRIPTION
**Pull request by hudson-ai**
JSON support for integer and float bounds.

Closes #1019 (thanks for the code @mmoskal!)

### CODE REVIEW

**Comment by hudson-ai** [comment]
Note: we have one xfail here due to the fact that JSON _apparently_ allows trailing `.00000...` on integers. We could add support for this in the future, but I don't really see a reason to (as doing so doesn't really add any expressive power to a model). Keeping the xfail around as a little piece of "documentation".

Ready for review.

**Comment by hudson-ai** [comment]
Another note: JSON Schema Draft4 has a slightly different convention around `exclusiveMinimum`/`exclusiveMaximum` (where they are treated as bools). We may want to allow schemas using this convention, just because I still see a lot of Draft4 in the wild.

Currently, our Draft202012 validator will reject these schemas, so we should be safe not to handle that case internally. Allowing Draft4 might open a bit of a can of worms, so I want to punt on that for at least a while.

**Comment by hudson-ai** [comment]
@riedgar-ms let me know if you're satisfied with the tests here :)

File: tests/unit/library/test_json.py [hunk 1]
```diff
@@ -194,6 +194,82 @@ def test_bad_number(self, bad_string, good_bytes, failure_byte, allowed_bytes, m
...
+    @pytest.mark.parametrize(
+        "instance, schema, should_pass",
+        [
+            # --- Integer type tests ---
+            (5, {"type": "integer", "minimum": 5}, True),
+            pytest.param(
+                *(5.0, {"type": "integer", "minimum": 5}, True),
+                marks=pytest.mark.xfail(reason="JSON technically allows trailing zeroes, but we currently don't")
+            ),
+            (5.1, {"type": "integer", "minimum": 5}, False),
```
Comment (riedgar-ms) [inline tests/unit/library/test_json.py:207]: "Integer" and "5.1" .... is this telling me something about JSON schema that I don't want to know?

File: tests/unit/library/test_json.py [hunk 2]
```diff
@@ -194,6 +194,82 @@ def test_bad_number(self, bad_string, good_bytes, failure_byte, allowed_bytes, m
...
+            (5, {"type": "number", "exclusiveMaximum": 5.0}, False),
+            (5.0, {"type": "number", "exclusiveMaximum": 5.0}, False),
+            (7.5, {"type": "number", "minimum": 5.0, "maximum": 10.0}, True),
+            (4.9, {"type": "number", "minimum": 5.0, "maximum": 10.0}, False),
+            (5.1, {"type": "number", "exclusiveMinimum": 5.0, "exclusiveMaximum": 10.0}, True),
+            (5.0, {"type": "number", "exclusiveMinimum": 5.0, "exclusiveMaximum": 10.0}, False),
+            (9.9, {"type": "number", "exclusiveMinimum": 5.0, "exclusiveMaximum": 10.0}, True),
+            # --- Edge cases ---
+            (0, {"type": "integer", "minimum": 0}, True),
+            (-1, {"type": "integer", "minimum": 0}, False),
```
Comment (riedgar-ms) [inline tests/unit/library/test_json.py:240]: Is there a reason that these are the only tests of negative numbers?

File: tests/unit/library/test_json.py [hunk 3]
```diff
@@ -194,6 +194,82 @@ def test_bad_number(self, bad_string, good_bytes, failure_byte, allowed_bytes, m
...
+            (0, {"type": "number", "maximum": 0}, True),
+            (5, {"type": "integer", "minimum": 5.5}, False),
+            (6, {"type": "integer", "minimum": 5.5}, True),
+            (5, {"type": "integer", "exclusiveMinimum": 5.5}, False),
+            (6, {"type": "integer", "exclusiveMinimum": 5.5}, True),
+            (5, {"type": "integer", "maximum": 5.5}, True),
+            (6, {"type": "integer", "maximum": 5.5}, False),
+            (5, {"type": "integer", "exclusiveMaximum": 5.5}, True),
+            (6, {"type": "integer", "exclusiveMaximum": 5.5}, False),
+            # --- Large numbers ---
```
Comment (riedgar-ms) [inline tests/unit/library/test_json.py:251]: We could add a bunch more large number and precision cases (as I think manically about how floating point can really Mess You Up), but I wouldn't worry right now.

File: tests/unit/library/test_regex_utils.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+def do_test_float_range(rx, left: float, right: float, left_inclusive: bool = True, right_inclusive: bool = True, left_none: bool = False, right_none: bool = False):
+    left_int = math.ceil(left)
+    right_int = math.floor(right)
+    if left_int < right_int:
+        if not left_inclusive and left_int == left:
+            left_int += 1
+        if not right_inclusive and right_int == right:
+            right_int -= 1
+        do_test_int_range(rx, left_int, right_int, left_none, right_none)
+    eps = 0.000001
```
Comment (riedgar-ms) [inline tests/unit/library/test_regex_utils.py:146]: This looks like a `float` epsilon, not a `double` one?

**Review by riedgar-ms — APPROVED** [review]
My main concern is the lack of negative number tests in `TestBoundedNumeric`.

File: tests/unit/library/test_regex_utils.py (hunk 4 above)
Comment (mmoskal) [reply tests/unit/library/test_regex_utils.py:146]: This was just supposed to be a small number. I think we could probably test with 0.0000000001 all way down to 0.1 each 10x or something. AFAIK JSON schema doesn't say numbers are doubles.

Comment (riedgar-ms) [reply tests/unit/library/test_regex_utils.py:146]: I wouldn't worry about that right now; my main concern is having more negative numbers in the main test cases. I have the impression that JSON may be a bit _relaxed_ in its concepts of floating point.

File: tests/unit/library/test_json.py (hunk 1 above)
Comment (hudson-ai) [reply tests/unit/library/test_json.py:207]: Don't worry, this is a "should fail" case (see the False in the third slot of the tuple 😄)

File: tests/unit/library/test_json.py (hunk 2 above)
Comment (hudson-ai) [reply tests/unit/library/test_json.py:240]: No, that was a terrible oversight on my part. There are a lot more test cases on the underlying regex_utils, but I'll make sure to get some populated in the higher-level json tests.

Comment (riedgar-ms) [reply tests/unit/library/test_json.py:240]: Just the core cases should be enough. Don't need to re-test everything exhaustively.

File: tests/unit/library/test_json.py (hunk 1 above)
Comment (riedgar-ms) [reply tests/unit/library/test_json.py:207]: Probably should have made the comment on the previous line :-) The 'reason' gives me bad feelings about potentially subtle bugs (but we have to follow the JSON standard, so 🤷)

**Comment by hudson-ai** [comment]
@riedgar-ms added some negative test cases. Will merge once I see some green :)
